de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ISarosSession.java

```diff
@@ -387,26 +387,17 @@ public void addSharedResources(
   public void changeColor(int colorID);
 
   /**
-   * FOR INTERNAL USE ONLY !
-   *
-   * <p>Starts queuing of incoming {@linkplain IResourceActivity project-related activities}, since
-   * they cannot be applied before their corresponding project is received and extracted.
+   * Register a queue handler for an incoming project negotiation. Should be called only once or
```

Comment:
Great now you glue the Negotiation stuff even harder into the session management. While we tried to make the old methods look like some general generic API methods this new description also tells you the only use case.

Also note there is no way you can "safely" change the handler as you have to do two calls. It does not matter on how often you call flushQueue as activities can arrive at any time and so it is possible that you replace a queuer that still has pending activities.

Reply:
Ok, I remove the remark to incoming project negotiation.

Regarding the replacement: you would have to ensure that the Queuer you want to replace has a empty queue, how you do it is none of my business (tell the Queuer to dequeue all on next process call and don't queue new ones; flush; check if empty).

Reply:
Have you read my comment ? It is NOT possible without hacking further into the session stuff. So either you add logic to make this possible or restrict the setQueuer call to only be allowed once during a session so you cannot replace it anymore.

The process method is called by a network thread, the offered session methods are called by some other threads.

Didn't you already noticed we added the queuing in such a way that you normally cannot control it ? The SarosSession is no playground for new developers as this is plagued by multiple thread access.

Reply:
I added a check to only allow one queue for a session